File: trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int left = 0;
        int right = height.length - 1;
        
        int leftMax = 0;
        int rightMax = 0;
        
        int count = 0;
        while (left < right) {
            if (height[left] < height[right]) {
                leftMax = Math.max(leftMax, height[left]);
                count += leftMax - height[left];
                left++;
            } else {
                rightMax = Math.max(rightMax, height[right]);
                count += rightMax - height[right];
                right--;
            }
        }
        
        return count;
    }
}

File: trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left_right_max = [[0, 0] for _ in range(len(height))]
        curr_max = 0
        for i in range(len(height)):
            h = height[i]
            curr_max = max(curr_max, h)
            left_right_max[i][0] = curr_max
        
        curr_max = 0
        for i in range(len(height) - 1, -1, -1):
            h = height[i]
            curr_max = max(curr_max, h)
            left_right_max[i][1] = curr_max
        
        ans = 0
        for i, (left_max, right_max) in enumerate(left_right_max):
            ans += min(left_max, right_max) - height[i]
            
        return ans
        
        